### PR TITLE
Use numpy cast functions to cast arrays

### LINKED ISSUE #809: Numpy.complex64 and Numpy.complex128 doesn’t work with arrays
**Describe the bug**
Pyccel can't generate Fortran/Python code when using Numpy.complex64 and Numpy.complex128 with arrays.

**To Reproduce**
Provide code to reproduce the behavior:
```python
@types('int[:,:]')
def get_complex64_arr_2d(arr):
    from numpy import complex64, shape
    a = complex64(arr)
    s = shape(a)
    return len(s), s[0], s[1], a[0,0], a[0,1]
```

The error message:
`AttributeError: _dtype`

**Expected behavior**

**Language**
Python/Fortran

**Additional context**
tests are already in #719.

### PR DESCRIPTION
**Pull request by EmilyBourne**
Python's builtin functions cannot be called with arrays as arguments. Internally pyccel's AST allows this in order to facilitate casting. However this sometimes leads to confusion. In python numpy also has casting functions which can be called on arrays and have the expected behaviour. This PR therefore restores the classes `PythonInt`, `PythonFloat`, `PythonComplex` and `PythonBool` to classes which only handle scalars, ensures that all Numpy casting functions handle arrays, and calls the numpy casting functions instead of the pure python casting functions. This fixes #809 

**Commit Summary**
- [BUGFIX] Correctly set dtype etc for empty tuples
- Ensure that all numpy casting functions accept arrays as arguments (fixes #809)
- Use numpy casting instead of python casting when casting internally

### CODE REVIEW

**Review by mostafaMamoni — APPROVED** [review]
i didn't notice that in my journey with numpy functions, we definitely shouldn’t change the behaviour of Python builtins even in Pyccel's internals.

File: pyccel/ast/numpyext.py [hunk 1]
```diff
@@ -171,6 +190,12 @@ class NumpyReal(PythonReal):
     1.0
     """
     __slots__ = ('_rank','_shape','_order')
+    def __new__(cls, arg):
+        if isinstance(arg.dtype, NativeBool):
+            return NumpyInt(arg)
```
Comment (yguclu) [inline pyccel/ast/numpyext.py:195]: Shouldn't we also handle the case where `isinstance(arg.dtype, NativeInt)`?

Comment (EmilyBourne) [reply pyccel/ast/numpyext.py:195]: This is handled in `PythonReal.__new__` which contains:
```
def __new__(cls, arg):
        if isinstance(arg.dtype, NativeBool):
            return PythonInt(arg)
        elif not isinstance(arg.dtype, NativeComplex):
            return arg
        else:
            return super().__new__(cls)
```
Here it is necessary to handle booleans as in numpy function calls the arguments can be arrays. For other types there is no casting involved, we just return the argument so the contents of `PythonReal.__new__` are sufficient. As a result I didn't think it was necessary to duplicate the condition.

Comment (yguclu) [reply pyccel/ast/numpyext.py:195]: I see, thanks.

**Review by yguclu — CHANGES_REQUESTED** [review]
Could you add a couple of tests checking that #809 is fixed?

**Comment by EmilyBourne** [comment]
> Could you add a couple of tests checking that #809 is fixed?

I knew I'd forgotten something ! I have activated (and fixed) the tests which were created in PR #719 . I ran into some unexpected python problems which required reordering classes in ast/numpyext.py. It is now working, but issue #802 is preventing me from adding the tests to our unit tests

File: tests/epyccel/recognised_functions/test_numpy_funcs.py [hunk 2]
```diff
@@ -54,11 +54,20 @@
...
+        else:
+            return isinstance(pyccel_result, type(python_result)) or \
+                    isinstance(pyccel_result.item(), type(python_result))
     else:
-        return isinstance(pyccel_result, type(python_result))
+        if isinstance(python_result, np.generic):
+            #TODO: Remove when #735 is fixed
+            return isinstance(pyccel_result, (type(python_result.item()), type(python_result)))
+        else:
+            return isinstance(pyccel_result, type(python_result))
```
Comment (yguclu) [inline tests/epyccel/recognised_functions/test_numpy_funcs.py:71]: It seems to me that when `isinstance(pyccel_result, type(python_result)) == True` this function always returns `True`. If that is correct, we can greatly simplify this code with an early exit condition:

```python
if type(pyccel_result) is type(python_result):
    return True
```

And then treat the other corner cases.

Comment (EmilyBourne) [reply tests/epyccel/recognised_functions/test_numpy_funcs.py]: You are right. Done in ac9de23

**Review by yguclu — APPROVED** [review]
Great job! Thank you!
